.github/workflows/check.yaml

```
@@ -178,6 +178,8 @@ jobs:
           fi
           if [[ "${{ matrix.juju-channel }}" =~ 3\.[0-9]+\/stable ]]; then
             echo "TEST_JUJU3=1" >> "$GITHUB_ENV"  # https://github.com/openstack-charmers/zaza/pull/653
+          else
+            echo "TEST_JUJU3=''" >> "$GITHUB_ENV"
```

This may be why the func tests are failing:

```suggestion
            echo "TEST_JUJU3=" >> "$GITHUB_ENV"
```

---

(It's currently set to `''`, which zaza may be interpreting as a non-empty string (ie. truthy), and installing juju3 support instead.)

---

> @samuelallan72 I've tried to change the TEST_JUJU3 to have empty string because 0 would install juju 3.

@gabrielcocenza please see above - I think the issue was that it wasn't being set to an empty string in CI - it was being set to a string containing `''`.

---

(I committed the suggestion; let's see if it works)

---

It did work :open_mouth: 
So before in the failed job, debug step show `TEST_JUJU3=''`, but now it's simply `TEST_JUJU3= ` 
That kinda :exploding_head: because it means zaza evaluates the `''` to be a non-empty string. I copy that to my local shell and python still interprets correctly as empty. Not sure it's due to python version, but I learn something new (or weird) today...

---

Let me try to clear up what's happening here. Note that there are two levels here:

## Python

> That kinda 🤯 because it means zaza evaluates the `''` to be a non-empty string.

That's because `''` is a non-empty string. An empty string is ``. From python's perspective here, it was receiving a string with the contents of `''`. Now it's receiving a real empty string. Try:

```text
>>> nonempty = "''"
>>> empty = ""
>>> print(nonempty)
''
>>> print(empty)

```

## Shell

>  I copy that to my local shell and python still interprets correctly as empty. 

That's because your shell interprets the single quotes in `TEST_JUJU3=''`. Eg. see the following:

```text
$ TEST_JUJU3=''
$ echo $TEST_JUJU3

$ TEST_JUJU3="''"
$ echo $TEST_JUJU3
''
```

Same as if you did:

```text
$ TEST='foo'
$ echo $TEST
foo
$ TEST=''
$ echo $TEST

```


The difference you're seeing between the CI and your local shell is because:

- GITHUB_ENV processing seems to simply interpret everything after the `=` as the string contents.
- bash/sh/zsh will interpret single/double quotes as per the shell quoting config.

---

Ah, I see... Yes, it does look like GITHUB_ENV has a different interpreting as you said, I didn't think of it before. 
Thanks for the detailed explanation!